### ISSUE TITLE
Discussion - Process Enrolment

### ISSUE THREAD

**Issue by BarisSari** [ACWIC/training-provider-admin#14]
As it's mentioned in [#102](https://github.com/monkeypants/transferrable_aged_care_microcredentials/issues/102), this ticket is for discussing filtering criteria for the process enrolment task. 

----
I'm starting with the documents. These are from **ARHITECTURE.md**:

```gherkin
 Feature: process_enrolments
      As a Training Provider
      I need to process enrolments as they arise
      So that Aged Care Providers
          can enrol employees into courses
```

![image](https://user-images.githubusercontent.com/41292272/98781974-458ae180-2408-11eb-8b32-93143cb71db1.png)

----
What Chris suggested in the ticket is as follows:

> We need a way for the training provider to get all enrolments that meet a certain criteria.
> 
>  for a given course
>  for a given date range (date of enrolment)
>  for a given received/update date
>  etc - maybe other filtering criteria later: we should have an open discussion ticket in the ACWIC/training-provider-admin repo
> 
> As an absolute minimum, the training-provider-admin needs a GET /enrolments endpoint! Then we can have GET /enrolments?{filter_key=value} features.

**Comment by BarisSari** [comment]
I'm actually having some troubles to understand the flow of the process. Training providers will process enrolments and they will enrol a student to a course. But when an enrolment request is sent, we should check some criteria to understand if it's possible or not (such as date is appropriate, course is not full, etc.). And @monkeypants, you suggest to have GET /enrolments request to check these criteria.  Is that correct?

**Comment by Sminney** [comment]
Overall Process is something like...

1. Training provider adds course to catalogue, includes features, mapping etc that will enable course to match criteria
2. Aged Care provider (ACP) connects to catalogue to search for courses 
3. ACP searches by some or all possible features 
1. ACP decides to enrol employee into course
5. Shared secret and enrol process commenced by ACP


The above is undertaken in the tickets mentioned in #102 

This ticket relates to the "process enrolment" on the right hand side of the diagram, ie the process of enrolling from the training providers admin perspective. 

The results of many of these events would mean the training provider is able to get all enrolments that match certain features.

Hope that assists

**Comment by BarisSari** [comment]
Thanks for the explanation, @Sminney. I have thought about it and then I checked the TP microservices. I think I have a clearer mind now. I have a few questions but I think they are not related to this ticket. So, I'll ask them in another place.

**Comment by BarisSari** [comment]
Hello @Sminney and @monkeypants. I opened a PR which adds some basic stuff: `GET /enrolments` endpoint, `Enrolment` entity & repository classes, an usecase named `ProcessEnrolments`. However, the usecase is a bit pointless at the moment. It looks like this:

https://github.com/ACWIC/training-provider-admin/blob/9039483e8e949dde0c005cb21713ab6f51011fe5/app/use_cases/process_enrolments.py#L21-L31

It takes a dictionary as a parameter, which has these attributes: `course_id`, `start_date` (optional), `end_date` (optional), `receive_data` (optional). It finds the course which ACP decides to enrol employee. If there is no course, a validation error is thrown. After that, it gets all enrolment objects using the dictionary. Then, I believe that we should have some checks/standards to accept/reject enrolments but there isn't anything in the code. I couldn't find anything in the documentation as well. That's why I'd like to ask to you @Sminney, @monkeypants. What should we have here?

One other question. If we reject an enrolment, should we delete the enrolment object or use soft delete by adding a field something like `deleted` to `Enrolment` domain and mark it True when it's rejected?

**Comment by Sminney** [comment]
Good question, the assumption has always been that the Training Provider will add courses to the catalogue, so that would be where the course_id will be found. So the Training Providers Catalogue holds (possibly creates) the course code, the code might have date or duration added to it to make it less generic.
If there is a code the Training Provider adds to the catalogue that can't be enrolled against, when the ACP goes to enrol not only should the ACP get a rejected message, but the Training Provider should probably get a message like "you said course_id was available in the catalogue, but you won't let enrolments occur, is the catalogue or your enrolment broken?

I'm wondering if using Faker (https://faker.readthedocs.io/en/master/index.html)or similar to get us set of starting data to populate the catalogue, etc might help tune this.

I think if there is a legitimate course_id that fails it is a problem at the training provider side.

**Comment by monkeypants** [comment]
Let's call it an "enrollment_request" to avoid confusion.

The training provider needs to be able to:
* reject an enrolment_request (with a given reason)
* accept an enrolment_request (... now it really is an enrolment)
* request more information (with a message)

Maybe other things too, but those should do for now.

So the enrolment_request has some states:
* new
* info_requested
* accepted
* rejected
* cancelled
* withdrawn
* completed

So processing an enrolment really means changing it's state. Not all state transitions make sense. These ones do:
* new -> accepted
* new -> rejected
* new -> info_requested
* info_requested -> accepted
* info_requested -> rejected
* info_requested -> info_requested (i.e. "more info requested)
* accepted -> cancelled
* accepted -> withdrawn
* accepted -> completed

How do you think those should happen? `PATCH /enrolment_requests/{enrolment_request_uid}`?

**Comment by BarisSari** [comment]
Yes, I think training providers can use `GET /enrolment_requests` (with filter by `state` option) to get all enrolment requests they want. Then, they can use `PATCH /enrolment_requests/{enrolment_request_uid}` to change its state. I'm updating my PR.
